doc/specs/#476 - Package Pinning.md

```diff
@@ -0,0 +1,205 @@
+  
+- Remove package pinning configuration:
+
+  `winget pin remove <package> [--source <source>] [--force]`
+  
+- List package pinning configuration: 
+  
+  `winget pin list <package> [--source <source>]` for a specific package or `winget pin list` to list all
+
+#### Blocking
+To block a package from upgrade, use `winget pin add <package> --blocking`
+```text
+cmd> winget pin Microsoft.TestApp --blocking
+```
+Now the pinning configuration is recorded as
+```text
+PackageIdentifier       SourceIdentifier       Version        PinningType
+----------------------------------------------------------------------------
+Microsoft.TestApp       winget                                Blocking
+Microsoft.TestAppStore  msstore                               Blocking
+```
+**Note:** by default packages correlated from all sources are blocked, user can pass in `--source` to block for a specific source
+
+Corresponding upgrade behavior
+```text
+cmd> winget upgrade -all
+Microsoft TestApp is blocked from upgrade and skipped
+
+cmd> winget upgrade Microsoft.TestApp
+Microsoft TestApp is blocked from upgrade
+
+cmd> winget upgrade Microsoft.TestApp --force
+Success
+```
+
+#### Pinning
+To pin a package from `winget upgrade --all`, use `winget pin add <package>`
+```text
+cmd> winget pin Microsoft.TestApp
+```
+Now the pinning configuration is recorded as
+```text
+PackageIdentifier       SourceIdentifier       Version        PinningType
+----------------------------------------------------------------------------
+Microsoft.TestApp       winget                                Pinning
+Microsoft.TestAppStore  msstore                               Pinning
+```
+**Note:** by default packages correlated from all sources are pinned, user can pass in `--source` to pin for a specific source
+
+Corresponding upgrade behavior
+```text
+cmd> winget upgrade -all
+Microsoft TestApp is pinned from upgrade and skipped
+
+cmd> winget upgrade Microsoft.TestApp
+Success
+```
+
+#### Gating
+To gate a package to some specific version, use `winget pin add <package> --version <gated version>`
+```text
+cmd> winget pin Microsoft.TestApp --version 1.2.*
+```
+Now the pinning configuration is recorded as
+```text
+PackageIdentifier       SourceIdentifier       Version        PinningType
+----------------------------------------------------------------------------
+Microsoft.TestApp       winget                 1.2.*          Gating
+Microsoft.TestAppStore  msstore                1.2.*          Gating
+```
+**Note:** by default packages correlated from all sources are gated, user can pass in `--source` to gate for a specific source
+
+Corresponding upgrade behavior
+```text
+cmd> winget upgrade -all
+Success  // If the available versions for upgrade are: 1.2.3 and 1.3.0, the selected version for upgrade is 1.2.3
+
+cmd> winget upgrade Microsoft.TestApp
+Success  // If the available versions for upgrade are: 1.2.3 and 1.3.0, the selected version for upgrade is 1.2.3
```

Comment:
What if the package gets updated to 1.3.0 outside of winget? Do we show a warning, offer a way to re-install the pinned version?

Reply:
I've commented about that situation before. I think it would be good for WinGet to show when a "pin" has been violated in cases other than "blocking". I think we would also want to make sure that if a user applied "--force" or some other formal override, that the pin is either removed or ideally updated and the user is informed.